File: mobile/lib/core/api/api_service.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:youtube_daily_digest/core/api/api_client.dart';
import 'package:youtube_daily_digest/features/topics/models/topic.dart';
import 'package:youtube_daily_digest/features/videos/models/video.dart';

final dioProvider = Provider<Dio>((ref) => createDioClient());

final apiServiceProvider = Provider<ApiService>((ref) {
  return ApiService(ref.watch(dioProvider));
});

class ApiService {
  final Dio _dio;

  ApiService(this._dio);

  // ── Topics ────────────────────────────────────────────────────────────────

  Future<List<Topic>> getTopics() async {
    final response = await _dio.get<List<dynamic>>('/topics');
    final data = response.data ?? [];
    return data
        .map((item) => Topic.fromJson(item as Map<String, dynamic>))
        .toList();
  }

  Future<Topic> createTopic(String name) async {
    final response = await _dio.post<Map<String, dynamic>>(
      '/topics',
      data: {'name': name},
    );
    return Topic.fromJson(response.data!);
  }

  Future<Map<String, dynamic>> runTopicSearch(int topicId) async {
    final response = await _dio.post<Map<String, dynamic>>(
      '/topics/$topicId/search',
    );
    return response.data!;
  }

  // ── Queries ───────────────────────────────────────────────────────────────

  Future<TopicQuery> createQuery({
    required int topicId,
    required String queryText,
    required int priority,
  }) async {
    final response = await _dio.post<Map<String, dynamic>>(
      '/topics/$topicId/queries',
      data: {'queryText': queryText, 'priority': priority},
    );
    return TopicQuery.fromJson(response.data!);
  }

  Future<TopicQuery> updateQuery({
    required int queryId,
    required String queryText,
    required int priority,
  }) async {
    final response = await _dio.put<Map<String, dynamic>>(
      '/queries/$queryId',
      data: {'queryText': queryText, 'priority': priority},
    );
    return TopicQuery.fromJson(response.data!);
  }

  Future<TopicQuery> toggleQuery({
    required int queryId,
    required bool isActive,
  }) async {
    final response = await _dio.patch<Map<String, dynamic>>(
      '/queries/$queryId/toggle',
      data: {'isActive': isActive ? 1 : 0},
    );
    return TopicQuery.fromJson(response.data!);
  }

  Future<void> deleteQuery(int queryId) async {
    await _dio.delete<void>('/queries/$queryId');
  }

  // ── Videos ────────────────────────────────────────────────────────────────

  Future<VideosResponse> getVideos({
    int page = 1,
    int pageSize = 50,
    String? date,
    String? search,
  }) async {
    final queryParams = <String, dynamic>{
      'page': page,
      'pageSize': pageSize,
    };
    if (date != null && date.isNotEmpty) queryParams['date'] = date;
    if (search != null && search.isNotEmpty) queryParams['search'] = search;

    final response = await _dio.get<Map<String, dynamic>>(
      '/videos',
      queryParameters: queryParams,
    );
    return VideosResponse.fromJson(response.data!);
  }
}

File: ui/src/app/services/api.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import type { Topic, TopicQuery, VideosResponse } from '../models';

@Injectable({ providedIn: 'root' })
export class ApiService {
  private readonly http = inject(HttpClient);
  private readonly base = '/api/v1';

  getTopics(): Observable<Topic[]> {
    return this.http.get<Topic[]>(`${this.base}/topics`);
  }

  createTopic(name: string): Observable<Omit<Topic, 'queries' | 'videoCount'>> {
    return this.http.post<Omit<Topic, 'queries' | 'videoCount'>>(`${this.base}/topics`, { name });
  }

  searchTopic(topicId: number): Observable<{ started: boolean; topicId: number }> {
    return this.http.post<{ started: boolean; topicId: number }>(
      `${this.base}/topics/${topicId}/search`,
      {},
    );
  }

  addQuery(topicId: number, queryText: string, priority: number): Observable<TopicQuery> {
    return this.http.post<TopicQuery>(`${this.base}/topics/${topicId}/queries`, {
      queryText,
      priority,
    });
  }

  updateQuery(queryId: number, queryText: string, priority: number): Observable<TopicQuery> {
    return this.http.put<TopicQuery>(`${this.base}/queries/${queryId}`, { queryText, priority });
  }

  deleteQuery(queryId: number): Observable<{ success: boolean }> {
    return this.http.delete<{ success: boolean }>(`${this.base}/queries/${queryId}`);
  }

  toggleQuery(queryId: number, isActive: number): Observable<TopicQuery> {
    return this.http.patch<TopicQuery>(`${this.base}/queries/${queryId}/toggle`, { isActive });
  }

  deleteVideo(videoId: string, reason: string): Observable<{ success: boolean; deletedVideoId: string }> {
    return this.http.delete<{ success: boolean; deletedVideoId: string }>(
      `${this.base}/videos/${encodeURIComponent(videoId)}`,
      { body: { reason } },
    );
  }

  getVideos(page = 1, pageSize = 50, date?: string, englishOnly = true): Observable<VideosResponse> {
    const params: Record<string, string> = {
      page: String(page),
      pageSize: String(pageSize),
      englishOnly: String(englishOnly),
    };
    if (date) params['date'] = date;
    return this.http.get<VideosResponse>(`${this.base}/videos`, { params });
  }
}
